go-controller/pkg/observability/observability.go

```diff
@@ -0,0 +1,210 @@
+package observability
+
+import (
+	"fmt"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	libovsdbops "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/libovsdb/ops"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+// OVN observ app IDs. Make sure to always add new apps in the end.
+const (
+	DropSamplingID = iota + 1
+	ACLNewTrafficSamplingID
+	ACLEstTrafficSamplingID
+)
+
+// temporary const, until we have dynamic config
+const DefaultObservabilityCollectorSetID = 28
+
+// this is inferred from nbdb schema, check Sample_Collector.id
+const maxCollectorID = 255
+
+// collectorConfig holds the configuration for a collector.
+// It is allowed to set different probabilities for every feature.
+// collectorSetID is used to set up sampling via OVSDB.
+type collectorConfig struct {
+	collectorSetID int
+	// probability in percent, 0 to 100
+	featuresProbability map[libovsdbops.SampleFeature]int
+}
+
+type Manager struct {
+	nbClient   libovsdbclient.Client
+	sampConfig *libovsdbops.SamplingConfig
+	// nbdb Collectors have probability. To allow different probabilities for different features,
+	// multiple nbdb Collectors will be created, one per probability.
+	// getCollectorKey() => collector.UUID
+	dbCollectors map[string]string
+	// Only maxCollectorID collectors are allowed, each should have unique ID.
+	// this set is tracking already assigned IDs.
+	takenCollectorIDs sets.Set[int]
+}
+
+func NewManager(nbClient libovsdbclient.Client) *Manager {
+	return &Manager{
+		nbClient:          nbClient,
+		dbCollectors:      make(map[string]string),
+		takenCollectorIDs: sets.New[int](),
+	}
+}
+
+func (m *Manager) SamplingConfig() *libovsdbops.SamplingConfig {
+	return m.sampConfig
+}
+
+func (m *Manager) Init() error {
+	if err := m.setSamplingAppIDs(); err != nil {
+		return err
+	}
+	if err := m.setDbCollectors(); err != nil {
+		return err
+	}
```

Comment:
What about potentially stale `NB.Sample_Collector` records, e.g. after upgrade some features have different probabilities?  AFAICT we don't delete any stale records in the DB.  Aside from the fact that those wouldn't be used for anything they also take up available IDs (and OVN currently only allows max 255 of those).

Reply:
that is not possible yet, as the only supported options are on/off without config change, but I have added some functionality to do that + unit tests (as it is not used by the real code yet)

Reply:
Thanks for adding tests!